src/Validator/Rules/UniqueFieldDefinitionNames.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Validator\Rules;
+
+use GraphQL\Error\Error;
+use GraphQL\Language\AST\InputObjectTypeDefinitionNode;
+use GraphQL\Language\AST\InputObjectTypeExtensionNode;
+use GraphQL\Language\AST\InterfaceTypeDefinitionNode;
+use GraphQL\Language\AST\InterfaceTypeExtensionNode;
+use GraphQL\Language\AST\NameNode;
+use GraphQL\Language\AST\NodeKind;
+use GraphQL\Language\AST\ObjectTypeDefinitionNode;
+use GraphQL\Language\AST\ObjectTypeExtensionNode;
+use GraphQL\Language\Visitor;
+use GraphQL\Language\VisitorOperation;
+use GraphQL\Type\Definition\InputObjectType;
+use GraphQL\Type\Definition\InterfaceType;
+use GraphQL\Type\Definition\NamedType;
+use GraphQL\Type\Definition\ObjectType;
+use GraphQL\Validator\SDLValidationContext;
+
+/**
+ * Unique field definition names.
+ *
+ * A GraphQL complex type is only valid if all its fields are uniquely named.
+ */
+class UniqueFieldDefinitionNames extends ValidationRule
+{
+    public function getSDLVisitor(SDLValidationContext $context): array
+    {
+        $schema = $context->getSchema();
+        /** @var array<string, NameNode[]> $knownFieldNames */
+        $knownFieldNames = [];
+        $checkFieldUniqueness = static function ($node) use ($context, $schema, &$knownFieldNames): VisitorOperation {
+            assert($node instanceof InputObjectTypeDefinitionNode || $node instanceof InterfaceTypeDefinitionNode || $node instanceof ObjectTypeDefinitionNode || $node instanceof InputObjectTypeExtensionNode || $node instanceof InterfaceTypeExtensionNode || $node instanceof ObjectTypeExtensionNode);
```

Comment:
No, there is no common interface. They are the 6 types of nodes supported by this rule: https://github.com/webonyx/graphql-php/pull/1111/files#diff-b20b47356cdad418fb8a83dd0e978923a68e02857888235e4b6fa9a03c5d303dR73-R78

Since the union can't be used as the parameter type and `$node` is just `mixed`, I added it here to get help from IDE and PHPStan while working on the code. And forgot to remove it afterwards. 

Unless you want me to keep it there, I will remove the `assert` before the PR is ready.  The problem with `$node` being `mixed` is common to all/most rules. We had a chat on this topic a while back: https://github.com/webonyx/graphql-php/pull/1111/files#diff-b20b47356cdad418fb8a83dd0e978923a68e02857888235e4b6fa9a03c5d303dR73-R78

Reply:
Got it, thanks for the explanation.

The intent behind my question was to see if we could define the code to be more self-explanatory that way. For example, if the assert were `assert($node instanceof NodeWithFields)`, it would be obvious why those node types were selected.

I am fine with this implementation and would like to keep the `assert()` - perhaps formatted to be split across multiple lines for readiblilty.